processing/src/main/java/io/druid/query/filter/LikeDimFilter.java

```diff
@@ -0,0 +1,358 @@
+
+      if (suffixPattern == null && suffixStyle == SuffixStyle.MATCH_PATTERN) {
+        throw new ISE("Need suffixPattern for pattern matching!");
+      } else if (suffixPattern != null && suffixStyle != SuffixStyle.MATCH_PATTERN) {
+        throw new ISE("Can't have suffixPattern without pattern matching!");
+      }
+    }
+
+    public static LikeMatcher from(
+        final String likePattern,
+        final char escapeChar,
+        final boolean extractPrefix
+    )
+    {
+      final StringBuilder prefix = new StringBuilder();
+      final StringBuilder regex = new StringBuilder();
+      boolean escaping = false;
+      boolean inPrefix = true;
+      for (int i = 0; i < likePattern.length(); i++) {
+        final char c = likePattern.charAt(i);
+        if (c == escapeChar && escapeChar != 0) {
+          escaping = true;
+        } else if (c == '%' && !escaping) {
+          inPrefix = false;
+          regex.append(".*");
+        } else if (c == '_' && !escaping) {
+          inPrefix = false;
+          regex.append(".");
+        } else {
+          if (inPrefix && extractPrefix) {
+            prefix.append(c);
+          } else {
+            // Excessively paranoid escaping, although shouldn't affect runtime beyond compiling the regex.
+            if (DEFINITELY_FINE.matcher(String.valueOf(c)).matches()) {
+              regex.append(c);
+            } else {
+              regex.append("\\u").append(BaseEncoding.base16().encode(Chars.toByteArray(c)));
+            }
+          }
+          escaping = false;
+        }
+      }
+
+      final String regexString = regex.toString();
+
+      if (regexString.isEmpty()) {
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_EMPTY, null);
+      } else if (regexString.equals(".*")) {
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_ANY, null);
+      } else {
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_PATTERN, Pattern.compile(regexString));
+      }
+    }
+
+    public boolean matches(@Nullable final String s)
+    {
+      if (s == null) {
+        return prefix == null && (suffixStyle == SuffixStyle.MATCH_ANY
+                                  || suffixStyle == SuffixStyle.MATCH_EMPTY
+                                  || suffixPattern.matcher("").matches());
+      }
+
+      final String suffix;
+
+      if (prefix == null) {
+        suffix = s;
+      } else if (s.startsWith(prefix)) {
+        suffix = s.substring(prefix.length());
+      } else {
+        return false;
+      }
+
+      if (suffixStyle == SuffixStyle.MATCH_ANY) {
+        return true;
+      } else if (suffixStyle == SuffixStyle.MATCH_EMPTY) {
+        return suffix.isEmpty();
+      } else {
+        // suffixStyle is MATCH_PATTERN
+        return suffixPattern.matcher(suffix).matches();
+      }
+    }
+
+    public DruidPredicateFactory predicateFactory(final ExtractionFn extractionFn)
+    {
+      return new DruidPredicateFactory()
```

Comment:
@gianm yes I mean local inner class. The name also serves as additional documentation sometimes, e. g. https://github.com/metamx/druid/commit/225c7e98c4a2116f78794ae3378b5a7f9a96bee7#diff-8da1c4dd87d76bcce224a6be6bf74ed8R615 (another extra LOC justifying using local inner class probably shouldn't be there if this would be a common practice)